### PR TITLE
[JENKINS-22367] Split JDKInstaller to a plugin

### PR DESCRIPTION
**Pull request by dwnusbaum**
See [JENKINS-22367](https://issues.jenkins-ci.org/browse/JENKINS-22367).

In #3147 I tried to remove `JDK.DescriptorImpl` as well, but ran into some [compiler errors](https://github.com/jenkinsci/jenkins/pull/3147#issuecomment-359575337) when preparing to release version 1.0 of the new plugin. I have created this PR starting from commit cbba7f0 to avoid unnecessary merges and reversions, but I am happy to update the original PR instead if desired.

~This PR currently just prepares the split by moving JDKInstaller and associated files to a module in the `jdk-tool-plugin` directory.~ The relevant files have been deleted here and moved to the [jdk-tool-plugin](https://github.com/jenkinsci/jdk-tool-plugin).

### Proposed changelog entries

* **Install from <code>java.sun.com</code>** installation method for JDK tools has been moved to a new _JDK Tool Plugin_.

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@reviewbybees

### CODE REVIEW

File: core/src/main/java/hudson/model/JDK.java [hunk 1]
```diff
@@ -183,8 +187,18 @@ public String getDisplayName() {
...
-            return Collections.singletonList(new JDKInstaller(null,false));
+        public List<? extends ToolInstaller> getDefaultInstallers() {
+            try {
+                Class<? extends ToolInstaller> jdkInstallerClass = Jenkins.getInstance().getPluginManager()
+                        .uberClassLoader.loadClass("hudson.tools.JDKInstaller").asSubclass(ToolInstaller.class);
+                Constructor<? extends ToolInstaller> constructor = jdkInstallerClass.getConstructor(String.class, boolean.class);
+                return Collections.singletonList(constructor.newInstance(null, false));
+            } catch (ClassNotFoundException e) {
+                return Collections.emptyList();
+            } catch (IllegalAccessException | InstantiationException | InvocationTargetException | NoSuchMethodException e) {
```
Comment (jglick) [inline core/src/main/java/hudson/model/JDK.java]: Or `ClassCastException` I suppose, theoretically. Maybe just catch `Exception` already.

File: jdk-tool-plugin/pom.xml [hunk 2]
```diff
@@ -0,0 +1,45 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.jenkins-ci.plugins</groupId>
+        <artifactId>plugin</artifactId>
+        <version>3.0</version>
```
Comment (jglick) [inline jdk-tool-plugin/pom.xml]: 3.5

File: jdk-tool-plugin/src/main/java/hudson/tools/JDKInstaller.java [hunk 3]
```diff
@@ -200,7 +208,7 @@ public void install(Launcher launcher, Platform p, FileSystem fs, TaskListener l
             }
 
             int exit = starter
-                    .stdin(new ByteArrayInputStream("yes".getBytes())).stdout(out)
+                    .stdin(new ByteArrayInputStream("yes".getBytes(Charset.defaultCharset()))).stdout(out)
```
Comment (jglick) [inline jdk-tool-plugin/src/main/java/hudson/tools/JDKInstaller.java]: `StandardCharsets.ASCII`?

File: jdk-tool-plugin/src/main/java/hudson/tools/JDKInstaller.java [hunk 4]
```diff
@@ -275,7 +283,7 @@ I tried to locate exactly how InstallShield parses the arguments (and why it use
                 out.println(Messages.JDKInstaller_FailedToInstallJDK(r));
                 // log file is in UTF-16
                 try (InputStreamReader in = new InputStreamReader(fs.read(logFile), "UTF-16")) {
-                    IOUtils.copy(in, new OutputStreamWriter(out));
+                    IOUtils.copy(in, new OutputStreamWriter(out, Charset.defaultCharset()));
```
Comment (jglick) [inline jdk-tool-plugin/src/main/java/hudson/tools/JDKInstaller.java]: `defaultCharset` on the master need not match the default charset on the agent.

File: jdk-tool-plugin/src/main/resources/hudson/tools/JDKInstaller/help.html [hunk 5]
```diff
@@ -0,0 +1,6 @@
+<div class="warning">
+    This installation method relies on a third-party service and may
+    break at any time without notice. Consider downloading the desired JDK
+    manually and then configuring the <q>Extract *.zip/*.tar.gz</q> installer
+    instead.
```
Comment (jglick) [inline jdk-tool-plugin/src/main/resources/hudson/tools/JDKInstaller/help.html]: Or simply preinstalling a JDK on agents, or using images with a JDK, etc.

File: jdk-tool-plugin/src/main/resources/org/jenkinsci/plugins/jdk_tool/Messages.properties [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+JDKInstaller.FailedToInstallJDK=Failed to install JDK. Exit code={0}
+JDKInstaller.RequireOracleAccount=Installing JDK requires Oracle account. <a href="{0}" target="_new">Please enter your username/password</a>
+JDKInstaller.UnableToInstallUntilLicenseAccepted=Unable to auto-install JDK until the license is accepted.
+JDKInstaller.DescriptorImpl.displayName=Install from java.sun.com
+JDKInstaller.DescriptorImpl.doCheckId=Define JDK ID
+JDKInstaller.DescriptorImpl.doCheckAcceptLicense=You must agree to the license to download the JDK.
```
Comment (jglick) [inline jdk-tool-plugin/src/main/resources/org/jenkinsci/plugins/jdk_tool/Messages.properties]: trailing newline

**Review by jglick — APPROVED** [review]
Sorry for the long delays here, I did not realize this was ready for final review.

File: jdk-tool-plugin/src/main/java/hudson/tools/JDKInstaller.java (hunk 3 above)
Comment (dwnusbaum) [reply jdk-tool-plugin/src/main/java/hudson/tools/JDKInstaller.java]: Yeah, not sure what I was thinking here.

File: jdk-tool-plugin/src/main/java/hudson/tools/JDKInstaller.java (hunk 4 above)
Comment (dwnusbaum) [reply jdk-tool-plugin/src/main/java/hudson/tools/JDKInstaller.java]: Ok, should I just pick some standard charset?

File: jdk-tool-plugin/src/main/resources/org/jenkinsci/plugins/jdk_tool/Messages.properties (hunk 6 above)
Comment (dwnusbaum) [reply jdk-tool-plugin/src/main/resources/org/jenkinsci/plugins/jdk_tool/Messages.properties]: Fixed in [d62e8dc](https://github.com/jenkinsci/jenkins/pull/3301/commits/d62e8dc0ce90f001bcc3d5b4bf05a64415a01d11#diff-c4c99b563e6a2043e20958e57d73823aR28).

File: jdk-tool-plugin/src/main/java/hudson/tools/JDKInstaller.java (hunk 4 above)
Comment (jglick) [reply jdk-tool-plugin/src/main/java/hudson/tools/JDKInstaller.java]: Do not know what is appropriate for this case. It is possible to obtain the agent’s charset if that is required (currently there is not a standard method to do so).

**Review by jglick — APPROVED** [review]

File: jdk-tool-plugin/src/main/java/hudson/tools/JDKInstaller.java [hunk 7]
```diff
@@ -284,7 +285,9 @@ I tried to locate exactly how InstallShield parses the arguments (and why it use
                 out.println(Messages.JDKInstaller_FailedToInstallJDK(r));
                 // log file is in UTF-16
                 try (InputStreamReader in = new InputStreamReader(fs.read(logFile), "UTF-16")) {
-                    IOUtils.copy(in, new OutputStreamWriter(out, Charset.defaultCharset()));
+                    Computer computer = Computer.currentComputer();
+                    Charset charset = computer != null ? computer.getDefaultCharset() : Charset.defaultCharset();
```
Comment (dwnusbaum) [inline jdk-tool-plugin/src/main/java/hudson/tools/JDKInstaller.java]: I don't think that `computer#getDefaultCharset` can return null here without a previous operation failing. I can add a null check after this line if that's not the case.

**Comment by dwnusbaum** [comment]
Pending [HOSTING-526](https://issues.jenkins-ci.org/browse/HOSTING-526).

**Comment by dwnusbaum** [comment]
@jglick I believe this is ready to go for the next weekly, can you take a final look? Maybe @daniel-beck is interested as well?

After the weekly is released I will release a `jdk-tool:1.1` after bumping the baseline to the newest weekly.

**Comment by daniel-beck** [comment]
I trust @jglick with &infinity; more experience than me with detaching plugins.

@dwnusbaum Just in case, did you did a manual test run on an existing instance to make sure the auto install works on upgrade?

**Comment by dwnusbaum** [comment]
@daniel-beck I just tested a 2.111 -> 2.112 upgrade and it worked correctly:

    INFO: Upgrading Jenkins. The last running version was 2.111. This Jenkins is version 2.112.
    Mar 16, 2018 10:20:58 AM hudson.PluginManager loadDetachedPlugins
    INFO: Upgraded Jenkins from version 2.111 to version 2.112. Loaded detached plugins (and dependencies): [jdk-tool.hpi]

**Review by jglick — APPROVED** [review]

**Comment by dwnusbaum** [comment]
@reviewbybees done

**Pull request merged**